File: js/arli/tests/run_tests.js
#!/usr/bin/env node

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { Evaluator } from '../src/eval.js';
import { arliRepr } from '../src/types.js';
import { tokenize, TokenStream } from '../src/tokenize.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const TEST_DIRS = [
  path.join(__dirname, '..', '..', '..', 'tests', 'common'),
  path.join(__dirname, '..', 'test_common'),
];

function findTestDir() {
  for (const d of TEST_DIRS) { if (fs.existsSync(d)) return d; }
  const projectTests = path.join(__dirname, '..', '..', '..', 'tests', 'common');
  if (fs.existsSync(projectTests)) {
    const localCopy = path.join(__dirname, '..', 'test_common');
    if (!fs.existsSync(localCopy)) {
      fs.mkdirSync(localCopy, { recursive: true });
      for (const f of fs.readdirSync(projectTests).filter(f => f.endsWith('.arli'))) fs.copyFileSync(path.join(projectTests, f), path.join(localCopy, f));
    }
    return localCopy;
  }
  return null;
}

function discoverTests(testDir) {
  return fs.readdirSync(testDir).filter(f => f.endsWith('.arli')).sort().map(f => ({ name: path.basename(f, '.arli'), filepath: path.join(testDir, f) }));
}

function parseExpected(filepath) {
  const expected = [];
  for (const line of fs.readFileSync(filepath, 'utf-8').split('\n')) {
    const m = line.match(/^\s*;;\s*expect:\s*(.*)/);
    if (m) expected.push(m[1].trim());
  }
  return expected;
}

function runJS(filepath) {
  const ev = new Evaluator();
  const output = [];
  const stream = new TokenStream(tokenize(fs.readFileSync(filepath, 'utf-8')));
  while (!stream.isEOF) {
    const expr = ev.parser._parseExpr(stream, true);
    if (expr !== null && expr !== undefined) {
      const result = ev.eval(expr);
      if (result !== null && result !== undefined) output.push(arliRepr(result));
    }
  }
  return output;
}

function runTest(name, filepath) {
  const expected = parseExpected(filepath);
  if (expected.length === 0) return { passed: true, expected: [], actual: [], errors: [] };
  const actual = runJS(filepath);
  const errors = [];
  let passed = true;
  const minLen = Math.min(expected.length, actual.length);
  for (let i = 0; i < minLen; i++) {
    if (expected[i] !== actual[i]) { errors.push(`  Line ${i + 1}: expected ${JSON.stringify(expected[i])}, got ${JSON.stringify(actual[i])}`); passed = false; }
  }
  for (let i = minLen; i < actual.length; i++) { errors.push(`  Extra output line ${i + 1}: ${JSON.stringify(actual[i])}`); passed = false; }
  for (let i = minLen; i < expected.length; i++) { errors.push(`  Missing line ${i + 1}: expected ${JSON.stringify(expected[i])}`); passed = false; }
  return { passed, expected, actual, errors };
}

function main() {
  const args = process.argv.slice(2);
  const showList = args.includes('--list');
  const filter = args.includes('--filter') ? args[args.indexOf('--filter') + 1] : '';
  const testDir = findTestDir();
  if (!testDir) { console.error('Error: Could not find test directory.'); process.exit(1); }
  const tests = discoverTests(testDir);

  if (showList) {
    console.log('Available tests:');
    for (const { name, filepath } of tests) console.log(`  ${name}: ${parseExpected(filepath).length} expectations`);
    return;
  }

  let passedCount = 0, failedCount = 0;
  for (const { name, filepath } of tests) {
    if (filter && !name.includes(filter)) continue;
    const { passed, errors } = runTest(name, filepath);
    console.log(`  [${passed ? 'PASS' : 'FAIL'}] ${name}`);
    if (passed) passedCount++;
    else { failedCount++; for (const err of errors.slice(0, 10)) console.log(`         ${err}`); if (errors.length > 10) console.log(`         ... and ${errors.length - 10} more errors`); }
  }

  const total = passedCount + failedCount;
  console.log(`\n${'='.repeat(50)}`);
  console.log(`  ${passedCount}/${total} passed, ${failedCount} failed\n  Backend: JavaScript`);
  process.exit(failedCount > 0 ? 1 : 0);
}

main();

File: tests/run_tests.py
#!/usr/bin/env python3
"""Common test runner for arli — runs .arli test files and verifies outputs.

Usage:
    python tests/run_tests.py              # test Python backend
    python tests/run_tests.py --go PATH    # test Go backend via binary
    python tests/run_tests.py --list       # list available tests
"""

import sys
import os
import subprocess
import re
import glob

# Add src to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))

TEST_DIR = os.path.join(os.path.dirname(__file__), 'common')
GO_BINARY = os.path.join(os.path.dirname(__file__), '..', 'go', 'arli', 'arli.exe')


# ---------------------------------------------------------------------------
# Test discovery
# ---------------------------------------------------------------------------

def discover_tests():
    """Return list of (test_name, filepath) tuples sorted by filename."""
    files = sorted(glob.glob(os.path.join(TEST_DIR, '*.arli')))
    tests = []
    for f in files:
        name = os.path.splitext(os.path.basename(f))[0]
        tests.append((name, f))
    return tests


def parse_expected(filepath):
    """Extract expected output lines from ;; expect: comments."""
    expected = []
    with open(filepath, 'r', encoding='utf-8') as f:
        for line in f:
            m = re.match(r'^\s*;;\s*expect:\s*(.*)', line)
            if m:
                expected.append(m.group(1).strip())
    return expected


# ---------------------------------------------------------------------------
# Python backend runner
# ---------------------------------------------------------------------------

def run_python(filepath):
    """Run a .arli file with the Python backend, return output lines."""
    from arli.eval import Evaluator
    from arli.types import arli_repr, nil

    ev = Evaluator()
    source = open(filepath, 'r', encoding='utf-8').read()

    # Capture output
    from io import StringIO
    import contextlib

    output = []
    tokens = None
    from arli.tokenize import tokenize
    from arli.parse import TokenStream

    tokens = tokenize(source)
    stream = TokenStream(tokens)
    while not stream.is_eof:
        expr = ev.parser._parse_expr(stream, allow_arity=True)
        if expr is not None:
            result = ev.eval(expr)
            if result is not None:
                output.append(arli_repr(result))
    return output


# ---------------------------------------------------------------------------
# Go backend runner
# ---------------------------------------------------------------------------

def run_go(filepath, go_binary=GO_BINARY):
    """Run a .arli file with the Go backend, return output lines."""
    if not os.path.exists(go_binary):
        return None  # Go binary not available

    result = subprocess.run(
        [go_binary, filepath],
        capture_output=True,
        text=True,
        timeout=30
    )
    if result.returncode != 0:
        return [f"<ERROR: {result.stderr.strip()}>"]

    output = [line.rstrip('\r') for line in result.stdout.split('\n') if line]
    return output


# ---------------------------------------------------------------------------
# Test runner
# ---------------------------------------------------------------------------

def run_test(name, filepath, use_go=False, go_binary=GO_BINARY):
    """Run a single test and return (passed, expected, actual, errors)."""
    expected = parse_expected(filepath)
    if not expected:
        return True, [], [], []  # No expectations to check

    if use_go:
        actual = run_go(filepath, go_binary)
        if actual is None:
            return False, expected, ["<Go binary not found>"], ["Go binary missing"]
    else:
        actual = run_python(filepath)

    errors = []
    passed = True

    # Compare line by line
    min_len = min(len(expected), len(actual))
    for i in range(min_len):
        if expected[i] != actual[i]:
            errors.append(f"  Line {i+1}: expected {expected[i]!r}, got {actual[i]!r}")
            passed = False

    # Extra lines in actual
    if len(actual) > len(expected):
        for i in range(len(expected), len(actual)):
            errors.append(f"  Extra output line {i+1}: {actual[i]!r}")
            passed = False

    # Missing lines in actual
    if len(actual) < len(expected):
        for i in range(len(actual), len(expected)):
            errors.append(f"  Missing line {i+1}: expected {expected[i]!r}")
            passed = False

    return passed, expected, actual, errors


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    import argparse

    parser = argparse.ArgumentParser(description='Run arli common tests')
    parser.add_argument('--go', nargs='?', const=GO_BINARY, default=None,
                        help='Test Go backend (optional path to binary)')
    parser.add_argument('--python', action='store_true',
                        help='Test Python backend (default)')
    parser.add_argument('--list', action='store_true',
                        help='List available tests')
    parser.add_argument('--filter', type=str, default='',
                        help='Run only tests matching this substring')
    args = parser.parse_args()

    use_go = args.go is not None
    go_binary = args.go if args.go else GO_BINARY

    if args.list:
        print("Available tests:")
        for name, path in discover_tests():
            expected = parse_expected(path)
            print(f"  {name}: {len(expected)} expectations")
        return

    if not args.python and not use_go:
        # Default: test Python
        args.python = True

    tests = discover_tests()
    passed_count = 0
    failed_count = 0

    for name, filepath in tests:
        if args.filter and args.filter not in name:
            continue

        passed, expected, actual, errors = run_test(
            name, filepath, use_go=use_go, go_binary=go_binary)

        result = "PASS" if passed else "FAIL"
        if passed:
            passed_count += 1
        else:
            failed_count += 1

        print(f"  [{result}] {name}")

        if not passed:
            for err in errors[:10]:  # Show first 10 errors
                print(f"         {err}")
            if len(errors) > 10:
                print(f"         ... and {len(errors) - 10} more errors")

    print(f"\n{'='*50}")
    total = passed_count + failed_count
    print(f"  {passed_count}/{total} passed, {failed_count} failed")
    print(f"  Backend: {'Go' if use_go else 'Python'}")

    return 0 if failed_count == 0 else 1


if __name__ == '__main__':
    sys.exit(main())
